### PR TITLE
Olivia & Ashton Scrabble Game

### PR DESCRIPTION
**Pull request by ashtn**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   How do you feel you and your partner did in sharing responsibilities?	|   Great! We basically spilt responsibilities one on code and one on test with overlap on refining and refactoring code. Maybe didn't follow a traditional driver and navigator. Immaculate white boarding skills!	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | Because the scoring class had a single job of calculating with methods that varied how things were calculated, all methods were called on score instead of an instance  |
|   Describe an example of a test you wrote for a _nominal case_. |   that the score method output an integer	|
|   Describe an example of a test you wrote for an _edge case_. |  passing non alpha characters into the play(word) method 	|
|   What was your final test coverage percentage?	|   BRB	|
|   Is there a specific piece of code you'd like feedback on?	|   refactoring feedback would be amazing for the highest_score_from method. More specifically how the code could be broken into smaller methods (if possible). This is a pattern that is hard to recognize.	|

Olivia was so great! <3

### CODE REVIEW

File: lib/scoring_class.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+
+      def self.highest_score_from(array_of_words)
+        raise ArgumentError.new("Must be an array!") if array_of_words.class != Array
+
+        hash_of_word_score = {}
+
+        array_of_words.each do |word|
+          hash_of_word_score[word] = self.score(word)
+        end
```
Comment (CheezItMan) [inline lib/scoring_class.rb:52]: There are really 3 logical steps in solving the problem.

Step 1:  Find each word's score

You could use `map` to get an array of each word's score here.

```
scored_words = array_of_words.map do |word|
  { score: score(word), word: word }
end
```
That gives you an array of hashes with words and their scores.  

Step 2:  Find the winning words
Then you can find the words with the highest score
```
highest_score = scord_words.max_by do |word_hash|
  word_hash[:score]
end

winners = scored_words.select do | word_hash|
  word_hash[:score] == highest_score
end
```

Step 3:  Pick the winner from the highest scoring words

If there's a 7-letter word return that, otherwise return the first word with the smallest length.

File: lib/scoring_class.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+              letter_points << num_key
+            end
+          end
+        end
+
+        letter_points.each do |point|
+          word_score += point
+        end
+
+        if word.length >= 7
```
Comment (CheezItMan) [inline lib/scoring_class.rb:36]: I think the max size word you can have in Scrabble is 7.  So it can't be greater than 7.

File: lib/scoring_class.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+        #not a tie
+        if array_of_highest_words.length == 1
+
+          #highest_scoring_word = array_of_highest_words.max_by {|k,v| v}
+          highest_scoring_word = array_of_highest_words.first
+
+          return highest_scoring_word
+
+          #instance of a tie #word lengths are the same
```
Comment (CheezItMan) [inline lib/scoring_class.rb:71]: What about a 7-letter word, in the case of a tie.

This if statement is also only checking two elements of the Array, the first & last element.

File: lib/player_class.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    def play(word)
+      raise ArgumentError.new("Only Letters A-Z") if (/\d|\W/) === word
+      @plays << word
+
+      tally = Scrabble::Scoring::score(word)
+
+      @total_score += tally
+
+      #puts @total_score #test
+      if won?
```
Comment (CheezItMan) [inline lib/player_class.rb:24]: You should add a word only if they haven't already won.

File: lib/player_class.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    # fills tiles array until it has 7 letters from the given tile bag
+    # def draw_tiles(tile_bag)
+    #     draw_tiles = 7 - @tiles
+    # end
+
+    def won?
+      if @total_score <= 100
+        false
+      else
+        "...and the winner is... #{@name}!! Game over bye :("
```
Comment (CheezItMan) [inline lib/player_class.rb:39]: This should be returning `true`, not a string

File: Specs/player_class_spec.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+   @ada.play(@word).must_equal false
+  end
+
+
+  it "Play(word) adds a word to @plays array" do
+
+    previous_length = @ada.plays.length
+    @ada.play("quizzed")
+    @ada.plays.length.must_be :>, previous_length
+  end
```
Comment (CheezItMan) [inline Specs/player_class_spec.rb:54]: You should also check that plays is empty to start.

File: lib/player_class.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      end
+    end
+
+    def highest_word_score(array_of_words)
```
Comment (CheezItMan) [inline lib/player_class.rb:43]: This shouldn't take an argument, it should use `@plays`

File: lib/player_class.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      word = Scoring::highest_score_from(array_of_words)
+      return Scoring::score(word)
+    end
+
+    def highest_scoring_word(array_of_words)
```
Comment (CheezItMan) [inline lib/player_class.rb:48]: ditto

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

|  Feature 	|   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions 	|   It looks like you've got quite a bit to work on with regards to testing	|
|   Both Teammates contributed to the codebase	|   Check 	|
|   Regular Commits with meaningful commit messages 	|   Good commit messages	|
|  **Scoring class**	|   Check	|
|  `score` method	|   A bit involved, but it works	|
|   Uses appropriate data structure to store the letter score	|   Good use of a hash to store letter scores.	|
|   Appropriately handles edge cases	|   Check	|
|   Tests `score` edge cases (one-letter word, 7 letter word)	|   Check	|
|  `highest_scoring_word` method	|   	|
|   Appropriately handles edge cases	|   It looks like you've got problems in your highest_scoring_word method.  You're having trouble getting the right answer when the word scores are tied.  I've got some notes in your PR code.	|
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|   You're skipping the `highest_scoring_word` tests...  however you are trying to test edge cases 	|
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|   Check	|
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|   Check, but the method should **not** take a parameter.  	|
|  Returns `true` or `false` from `won?` method |  The method is returning a String if they've won, which is "truthy," but not what was required.  |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |  The tests are skipped, but you're trying to test edge cases   |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|   Good use of an Array	|
|  `draw_tiles` method uses the parameter to draw a set of random tiles |   Sorta, but the Array is in order and you're popping a random # of elements.  It would be better to call `.shuffle` in `initialize` on your array and then pop the correct number of items.	|
| `tiles_remaining` method returns a count	|   Should be returning `@default_set.length`	|
|  Tests `draw_tiles` edge cases for different parameter inputs	|   Test file not completed	|

## Summary

It looks like you two had a lot of trouble with this project.  I would very much recommend you take advantage of the scheduled tutoring to learn from the project.  You can also schedule time to meet me before class and go over things.  Make sure you're taking advantage of TAs as well.  

Just a note, it's standard to name your files with the name of the class with underscores between words and in lower case.

**Review by CheezItMan — COMMENTED** [review]
Feedback in the code included.
